distributed/spill.py

```diff
@@ -1,76 +1,261 @@
         """Key/value pairs spilled out to disk. Alias of zict.Buffer.slow.
         For inspection only - do not modify directly!
         """
         return self.slow
 
-    @staticmethod
-    def _weight(key: Hashable, value: Any) -> int:
-        return safe_sizeof(value)
-
-    def _on_evict(self, key: Hashable, value: Any) -> None:
-        b = safe_sizeof(value)
-        self.spilled_by_key[key] = b
-        self.spilled_total += b
-
-    def _on_retrieve(self, key: Hashable, value: Any) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key)
-
-    def __setitem__(self, key: Hashable, value: Any) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key, 0)
-        super().__setitem__(key, value)
-        if key in self.slow:
-            # value is individually larger than target so it went directly to slow.
-            # _on_evict was not called.
-            b = safe_sizeof(value)
-            self.spilled_by_key[key] = b
-            self.spilled_total += b
-
-    def __delitem__(self, key: Hashable) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key, 0)
+    @property
+    def spilled_total(self) -> int:
+        """Number of bytes spilled to disk.
+        Note that this is the pickled size, which may differ from the output of sizeof().
+        """
+        return self.slow.total_weight
+
+
+def _in_memory_weight(key: str, value: Any) -> int:
+    return safe_sizeof(value)
+
+
+# Internal exceptions. These are never raised by SpillBuffer.
+class MaxSpillExceeded(Exception):
+    pass
+
+
+class PickleError(Exception):
+    pass
+
+
+class HandledError(Exception):
+    pass
+
+
+class Slow(zict.Func):
+    max_weight: int | Literal[False]
+    weight_by_key: dict[str, int]
+    total_weight: int
+
+    def __init__(self, spill_directory: str, max_weight: int | Literal[False] = False):
+        super().__init__(
+            partial(serialize_bytelist, on_error="raise"),
+            deserialize_bytes,
+            zict.File(spill_directory),
+        )
+        self.max_weight = max_weight
+        self.weight_by_key = {}
+        self.total_weight = 0
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        try:
+            pickled = self.dump(value)
+        except Exception as e:
+            # zict.LRU ensures that the key remains in fast if we raise.
+            # Wrap the exception so that it's recognizable by SpillBuffer,
+            # which will then unwrap it.
+            raise PickleError(key, e)
+
+        pickled_size = sum(len(frame) for frame in pickled)
+
+        if has_zict_210:
+            # Thanks to Buffer.__setitem__, we never update existing
+            # keys in slow, but always delete them and reinsert them.
+            assert key not in self.d
+            assert key not in self.weight_by_key
+        else:
+            self.d.pop(key, 0)
+            self.total_weight -= self.weight_by_key.pop(key, PickledSize(0, 0))
```

Comment:
@crusaderky it seems this is breaking the linting, what is PickledSize in here, it's not defined?

Reply:
my bad, it's from #5805